### PR TITLE
Screen3 changes

### PR DESCRIPTION
**Pull request by Vanshbajaj**
Feat - Added Ui State  in Presentation Layer
 
Handle the Expectations in Data Layer 
When the Internet is not connected
Internet is Connected

https://github.com/user-attachments/assets/c54bc55f-72c6-4693-a2e3-6ebc45b0bca9

### CODE REVIEW

File: data/src/test/java/com/practical/data/CharacterRepositoryImplTest.kt [hunk 1]
```diff
@@ -113,45 +112,128 @@ class CharacterRepositoryImplTest {
...
+            cancelAndIgnoreRemainingEvents()
+        }
+
+
+    }
+
 }
+
+
+
```
Comment (shreyansh-mm) [inline data/src/test/java/com/practical/data/CharacterRepositoryImplTest.kt:239]: Lets remove the unnecessary spacing

**Review by shreyansh-mm — CHANGES_REQUESTED** [review]

File: data/src/main/java/com/practical/data/network/AppException.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.practical.data.network
+
+
+open class NetworkException : Exception()
+
+class ClientNetworkException : NetworkException()
+
+class ApolloClientException : NetworkException()
```
[the remark this answers is not in the document]
Comment (Vanshbajaj) [reply data/src/main/java/com/practical/data/network/AppException.kt:8]: singleton class cannot be extended

File: data/src/main/java/com/practical/data/network/NetworkException.kt
Comment (shreyansh-mm) [inline data/src/main/java/com/practical/data/network/NetworkException.kt:1]: A lot of extra lines.

File: data/src/main/java/com/practical/data/repository/CharacterRepositoryImpl.kt [hunk 3]
```diff
@@ -21,58 +20,72 @@ class CharacterRepositoryImpl @Inject constructor(
...
+    override fun getCharactersList(): Flow<List<CharactersListModel>> {
+        return flow {
+            val response = apolloClient.query(CharactersListQuery()).execute()
+            val characters = response.data?.characters?.results?.mapNotNull { character ->
+                CharactersListModel(
+                    id = character?.id,
+                    name = character?.name,
+                    image = character?.image,
+                )
+            } ?: emptyList()
```
Comment (shreyansh-mm) [inline data/src/main/java/com/practical/data/repository/CharacterRepositoryImpl.kt:32]: Why are we emitting empty list?

File: presentation/src/main/java/com/practical/presentation/screens/CharacterScreen.kt [hunk 4]
```diff
@@ -41,47 +42,61 @@ fun CharacterScreen(
...
-                    CircularProgressIndicator(
-                        Modifier.align(
-                            Alignment.CenterVertically
+    state: UiState<CharacterModel>,
+    modifier: Modifier = Modifier,
+) {
+    Box(
+       modifier.fillMaxSize() // Fill the entire screen
+    ) {
+        Column{
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/practical/presentation/screens/CharacterScreen.kt:59]: Why do we need a `Box` and `Column` both?

**Review by shreyansh-mm — CHANGES_REQUESTED** [review]

**Review by shreyansh-mm — CHANGES_REQUESTED** [review]

File: data/src/main/java/com/practical/data/repository/CharacterRepositoryImpl.kt [hunk 5]
```diff
@@ -21,58 +20,47 @@ class CharacterRepositoryImpl @Inject constructor(
...
+                )
+            }
+            if (characters == null) {
+                throw NetworkException.ClientNetworkException
+            } else {
+                emit(characters)
+            }
+        }.catch {
+            throw NetworkException.ApolloClientException
         }
```
Comment (shreyansh-mm) [inline data/src/main/java/com/practical/data/repository/CharacterRepositoryImpl.kt:40]: catch here is not needed

File: presentation/src/main/java/com/practical/presentation/screens/CharactersListScreen.kt [hunk 6]
```diff
@@ -24,62 +25,81 @@ import androidx.compose.ui.graphics.Color
...
-                CharacterGrid(state.data, onNavigateToCharacterScreen)
-            }
+                        is NetworkException.ApolloClientException -> {
+                            Text(
+                                modifier = Modifier.align(Alignment.CenterHorizontally),
+                                text = stringResource(R.string.graphql_error),
+                                color = MaterialTheme.colorScheme.error
+                            )
+                        }
+                    }
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/practical/presentation/screens/CharactersListScreen.kt:87]: same, you can reuse the composable made for above screen

File: presentation/src/main/java/com/practical/presentation/viewmodel/CharacterViewModel.kt [hunk 7]
```diff
@@ -29,9 +30,29 @@ class CharacterViewModel @Inject constructor(
...
+                        is NetworkException.ApolloClientException -> {
+                            _charactersState.emit(UiState.Error(NetworkException.ApolloClientException))
+                        }
+
+                        else -> {
+                            _charactersState.emit(UiState.Error(it))
+                        }
+                    }
+
+                }
```
Comment (shreyansh-mm) [inline presentation/src/main/java/com/practical/presentation/viewmodel/CharacterViewModel.kt:48]: This logic can be improved

**Review by shreyansh-mm — CHANGES_REQUESTED** [review]

**Review by shreyansh-mm — APPROVED** [review]
Migrate common composable out of CharacterSceen to its own file.

**Pull request merged**